src/inmanta/agent/in_process_executor.py

```diff
@@ -67,12 +67,29 @@ def __init__(
         self.thread_pool: ThreadPoolExecutor = ThreadPoolExecutor(1, thread_name_prefix="Pool_%s" % self.name)
 
         self._cache = inmanta.agent.cache.AgentCache(self)
+        # This lock ensures cache entry can not be cleaned up when
+        # the executor is actively working and vice versa
+        self.wip_lock = asyncio.Lock()
 
         self.logger: logging.Logger = parent_logger.getChild(self.name)
 
         self._stopped = False
 
         self.failed_resource_types: FailedResourcesSet = set()
+        self.periodic_cache_cleanup_job: Optional[asyncio.Task[None]] = None
+
+    async def start(self) -> None:
+        self.periodic_cache_cleanup_job = asyncio.create_task(self.cleanup_stale_cache_entries())
+
+    async def cleanup_stale_cache_entries(self) -> None:
+        """
+        This task periodically cleans up stale entries in the cache
+        """
+        while True:
+            reschedule_interval: float = 1.0
+            async with self.wip_lock:
+                self._cache.clean_stale_entries()
+            await asyncio.sleep(reschedule_interval)
```

Comment:
A few lines below, the `self._cache.close()` should probably go on the threadpool as well  (requiring stop to be an async method), because it is called from the ioloop and it calls the finalizers